File: doc-examples/example-jsonb-java/src/main/java/example/ProgrammaticCodeDeserializer.java
package example;

import jakarta.json.bind.serializer.DeserializationContext;
import jakarta.json.bind.serializer.JsonbDeserializer;
import jakarta.json.stream.JsonParser;

import java.lang.reflect.Type;

public final class ProgrammaticCodeDeserializer implements JsonbDeserializer<ProgrammaticCode> {
    @Override
    public ProgrammaticCode deserialize(JsonParser parser, DeserializationContext ctx, Type rtType) {
        while (parser.hasNext()) {
            if (parser.next() == JsonParser.Event.VALUE_STRING) {
                return new ProgrammaticCode(parser.getString().substring(5));
            }
        }
        throw new IllegalStateException("Expected a JSON string");
    }
}

File: doc-examples/example-jsonb-kotlin/src/main/kotlin/example/ProgrammaticCodeDeserializer.kt
package example

import jakarta.json.bind.serializer.DeserializationContext
import jakarta.json.bind.serializer.JsonbDeserializer
import jakarta.json.stream.JsonParser
import java.lang.reflect.Type

class ProgrammaticCodeDeserializer : JsonbDeserializer<ProgrammaticCode> {
    override fun deserialize(parser: JsonParser, ctx: DeserializationContext, rtType: Type): ProgrammaticCode {
        while (parser.hasNext()) {
            if (parser.next() == JsonParser.Event.VALUE_STRING) {
                return ProgrammaticCode(parser.string.substring(5))
            }
        }
        throw IllegalStateException("Expected a JSON string")
    }
}
